File: packages/contracts/src/v1/phase-run.ts
import { z } from "zod";

import { PhaseDefinitionV1Schema, PhaseIdSchema, PhasePresetIdSchema } from "./phase.js";
import {
  GitBranchNameSchema,
  GitObjectIdSchema,
  IsoInstantSchema,
  NonNegativeSafeIntegerSchema,
  PhaseRunIdSchema,
  ProjectIdSchema,
  SchemaVersionV1Schema,
  Sha256DigestSchema,
} from "./primitives.js";
import { RoomIdSchema } from "./room.js";

/**
 * Phase Runner: `PhaseRunV1` is the durable, attempt-shaped record of one execution of a
 * `PhaseDefinitionV1` (`phase.run`/`phase.status`/`phase.list`/`phase.approve`/`phase.reject`).
 * It mirrors `ExecutionAttemptV1` (`execution.ts`) deliberately: a run is a state machine, not a
 * revisioned document, so — unlike `PhaseDefinitionV1`/`PhasePresetV1` — it has no `revision`-CAS
 * upsert; a run's `revision` field exists only for optimistic-concurrency writes as it advances
 * through its own state machine (`phase-run-state-machine.ts` in `@app-factory/kernel`).
 *
 * A run is bound immutably to the *exact bytes* of the phase definition it executes: `phaseSnapshot`
 * embeds the whole `PhaseDefinitionV1` value as read at creation time, and `phaseSnapshotDigest` is
 * the SHA-256 of its canonical JSON (`computePhaseDefinitionDigest`, `@app-factory/kernel`'s
 * `canonical-json.ts`). A later edit to the live phase definition (a new `phase.upsert` revision)
 * never changes a past run's snapshot or digest.
 *
 * `outputs[].evidence` deliberately does **not** reuse `EvidenceV1`/`EvidenceSubjectV1`
 * (`evidence.ts`): that envelope's subject shape (`taskSpecDigest`, `policyDigest`, `baseCommit`,
 * `candidateTree`, `fence`) is intrinsically TaskSpec/attempt-shaped and has no honest values for a
 * phase run (a phase run binds no TaskSpec, no policy decision, no scheduler fence). Forcing a phase
 * output through that envelope would mean inventing fake values for fields that do not apply, which
 * this repository's own discipline forbids (`rule.data.no-fake-fallback`). Instead, each output
 * carries its own direct commit-binding record (`PhaseRunOutputEvidenceV1`) — the same
 * `{commit, tree}` binding `CommitEvidenceV1.claims` (`evidence.ts`) carries for attempts, plus the
 * `factory/phase/<phaseId>/<runId>` branch it landed on — proof the content is a real, addressable
 * commit in the project's enrolled mirror, without borrowing a subject shape that does not fit.
 */

export const PhaseRunStateV1Schema = z.enum([
  "queued",
  "running",
  "awaiting-human",
  "succeeded",
  "failed",
  "cancelled",
]);
export type PhaseRunStateV1 = z.infer<typeof PhaseRunStateV1Schema>;

/** Mirrors `AttemptOutcomeV1` (`execution.ts`) exactly: the terminal-state payload. */
export const PhaseRunOutcomeV1Schema = z.discriminatedUnion("kind", [
  z.strictObject({ kind: z.literal("succeeded") }),
  z.strictObject({
    kind: z.literal("failed"),
    code: z.enum([
      "participant-error",
      "grader-changes-required",
      "output-schema-invalid",
      "output-commit-rejected",
      "rejected",
    ]),
    summary: z.string().min(1).max(2_000),
  }),
  z.strictObject({ kind: z.literal("cancelled"), reason: z.string().min(1).max(1_000) }),
]);
export type PhaseRunOutcomeV1 = z.infer<typeof PhaseRunOutcomeV1Schema>;

/**
 * Direct commit-binding proof for one committed output: the broker commit and its tree in the
 * project's enrolled Factory mirror, and the `factory/phase/<phaseId>/<runId>` branch it was
 * published on. See the module doc comment for why this does not reuse `EvidenceV1`.
 */
export const PhaseRunOutputEvidenceV1Schema = z.strictObject({
  commit: GitObjectIdSchema,
  tree: GitObjectIdSchema,
  branch: GitBranchNameSchema,
});
export type PhaseRunOutputEvidenceV1 = z.infer<typeof PhaseRunOutputEvidenceV1Schema>;

/**
 * The exact repo-relative path one committed output actually landed at in the project's mirror.
 * Deliberately its OWN schema rather than reusing `PhaseOutputV1`'s `path` verbatim (Seam (c) of the
 * project-registry task): a `PhaseDefinitionV1`'s declared `outputs[].path` is a project-agnostic
 * preset template that always names the canonical lowercase `docs/` convention, but the project a
 * run actually executes against may be registered under the capitalized `Docs/` convention instead
 * (`ProjectRegistryDocsDirV1Schema`) -- and on a case-insensitive storage volume, a write to the
 * declared `docs/...` path can genuinely land at `Docs/...` in the committed tree (APFS folds the
 * two, Git does not). This path must name where the file REALLY is, so evidence built from it (a
 * `git show <tree>:<path>` read-back) actually resolves -- reporting the declared, possibly-wrong-
 * case path here would silently produce evidence nothing can read back.
 */
export const PhaseRunOutputPathV1Schema = z
  .string()
  .min(1)
  .max(1_024)
  .refine(
    (value) => (value.startsWith("docs/") || value.startsWith("Docs/")) && !value.includes(".."),
    "output path must be repo-relative under the project's registered docs directory (docs/ or Docs/)",
  );

/** One committed output. */
export const PhaseRunOutputV1Schema = z.strictObject({
  path: PhaseRunOutputPathV1Schema,
  digest: Sha256DigestSchema,
  evidence: PhaseRunOutputEvidenceV1Schema,
});
export type PhaseRunOutputV1 = z.infer<typeof PhaseRunOutputV1Schema>;

/**
 * Deliberately lighter than `ReviewReportV1`/`FindingV1Schema`'s full code-review shape (no
 * `findingId`/`ruleId`/`category`/`severity`/`locations` — a phase grader judges prose outputs
 * against `rules.acceptanceChecks`, not source-code diffs, so findings are plain bounded text.
 * `verdict` is only `pass|changes-required` (never `blocked`): a phase grader either accepts the
 * produced outputs or does not — there is no "grader itself could not run" state distinct from a
 * `participant-error` run failure.
 */
export const PhaseRunGraderVerdictV1Schema = z.strictObject({
  verdict: z.enum(["pass", "changes-required"]),
  findings: z.array(z.string().min(1).max(2_000)).max(50),
});
export type PhaseRunGraderVerdictV1 = z.infer<typeof PhaseRunGraderVerdictV1Schema>;

export const PhaseRunTokenUsageV1Schema = z.strictObject({
  totalTokens: NonNegativeSafeIntegerSchema,
});
export type PhaseRunTokenUsageV1 = z.infer<typeof PhaseRunTokenUsageV1Schema>;

export const PhaseRunV1Schema = z.strictObject({
  schemaVersion: SchemaVersionV1Schema,
  phaseRunId: PhaseRunIdSchema,
  /** The preset this run was launched through, when one was named; `null` for a standalone phase. */
  presetId: PhasePresetIdSchema.nullable(),
  phaseId: PhaseIdSchema,
  projectId: ProjectIdSchema,
  phaseSnapshotDigest: Sha256DigestSchema,
  phaseSnapshot: PhaseDefinitionV1Schema,
  state: PhaseRunStateV1Schema,
  /** Optimistic-concurrency token for this run's own state-machine writes (not a document revision). */
  revision: NonNegativeSafeIntegerSchema,
  /** The persistent room a `chat`-mode run is bound to; always `null` for every other mode. */
  roomId: RoomIdSchema.nullable(),
  outputs: z.array(PhaseRunOutputV1Schema).max(10),
  graderVerdict: PhaseRunGraderVerdictV1Schema.nullable(),
  tokenUsage: PhaseRunTokenUsageV1Schema,
  outcome: PhaseRunOutcomeV1Schema.nullable(),
  createdAt: IsoInstantSchema,
  startedAt: IsoInstantSchema.nullable(),
  finishedAt: IsoInstantSchema.nullable(),
  updatedAt: IsoInstantSchema,
});
export type PhaseRunV1 = z.infer<typeof PhaseRunV1Schema>;

/** The wire input for `phase.run`: which phase to run, for which project, from which preset (if any). */
export const PhaseRunCreateV1Schema = z.strictObject({
  presetId: PhasePresetIdSchema.nullable(),
  phaseId: PhaseIdSchema,
  projectId: ProjectIdSchema,
  /** Overrides the phase definition's own declared `inputs[]` for this run only, when non-null. */
  inputsOverride: PhaseDefinitionV1Schema.shape.inputs.nullable(),
});
export type PhaseRunCreateV1 = z.infer<typeof PhaseRunCreateV1Schema>;

export const MAX_PHASE_RUN_LIST_ITEMS_V1 = 100 as const;

export const PhaseRunListCursorV1Schema = z.strictObject({
  updatedAt: IsoInstantSchema,
  phaseRunId: PhaseRunIdSchema,
});
export type PhaseRunListCursorV1 = z.infer<typeof PhaseRunListCursorV1Schema>;

export const PhaseRunListQueryV1Schema = z.strictObject({
  projectId: ProjectIdSchema.nullable(),
  state: PhaseRunStateV1Schema.nullable(),
  after: PhaseRunListCursorV1Schema.nullable(),
  limit: z.number().int().min(1).max(MAX_PHASE_RUN_LIST_ITEMS_V1),
});
export type PhaseRunListQueryV1 = z.infer<typeof PhaseRunListQueryV1Schema>;

export const PhaseRunListPageV1Schema = z
  .strictObject({
    runs: z.array(PhaseRunV1Schema).max(MAX_PHASE_RUN_LIST_ITEMS_V1),
    nextAfter: PhaseRunListCursorV1Schema.nullable(),
    hasMore: z.boolean(),
  })
  .superRefine((page, context) => {
    if (page.hasMore !== (page.nextAfter !== null)) {
      context.addIssue({
        code: "custom",
        path: ["nextAfter"],
        message: "nextAfter must be present exactly when hasMore is true",
      });
    }
    const identities = page.runs.map((run) => run.phaseRunId);
    if (new Set(identities).size !== identities.length) {
      context.addIssue({ code: "custom", path: ["runs"], message: "run IDs must be unique" });
    }
    if (page.nextAfter !== null) {
      const last = page.runs.at(-1);
      if (
        last === undefined ||
        last.updatedAt !== page.nextAfter.updatedAt ||
        last.phaseRunId !== page.nextAfter.phaseRunId
      ) {
        context.addIssue({
          code: "custom",
          path: ["nextAfter"],
          message: "nextAfter must identify the final returned run",
        });
      }
    }
  });
export type PhaseRunListPageV1 = z.infer<typeof PhaseRunListPageV1Schema>;

/** `phase.approve`/`phase.reject` payload: which run, and why (reject only; approve reason is optional). */
export const PhaseRunDecisionV1Schema = z.strictObject({
  phaseRunId: PhaseRunIdSchema,
  reason: z.string().min(1).max(1_000).nullable(),
});
export type PhaseRunDecisionV1 = z.infer<typeof PhaseRunDecisionV1Schema>;

File: apps/studio-mac/Sources/StudioKit/Models/PhaseRun.swift
import Foundation

// MARK: - Phase Runner (phase-run.ts, `phase.run`/`phase.status`/`phase.list`/`phase.approve`/
// `phase.reject`) — the one place a `PhaseDefinition` actually executes.
//
// `PhaseRun` mirrors `ExecutionAttempt`'s state-machine shape deliberately: a run is a state
// machine, not a revisioned document — its `revision` is an optimistic-concurrency token for the
// state machine's own writes, not a document-CAS revision like `PhaseDefinition.revision`. A run is
// bound immutably to the exact bytes of the phase it executes (`phaseSnapshot`/`phaseSnapshotDigest`
// embed the whole definition as read at creation time); a later edit to the live phase never changes
// a past run.

public enum PhaseRunState: String, Hashable, Sendable, Codable, CaseIterable {
    case queued, running
    case awaitingHuman = "awaiting-human"
    case succeeded, failed, cancelled
}

/// Mirrors `AttemptOutcome`'s role: the terminal-state payload.
public enum PhaseRunOutcome: Hashable, Sendable {
    case succeeded
    case failed(code: PhaseRunFailureCode, summary: String)
    case cancelled(reason: String)
}

public enum PhaseRunFailureCode: String, Hashable, Sendable, Codable, CaseIterable {
    case participantError = "participant-error"
    case graderChangesRequired = "grader-changes-required"
    case outputSchemaInvalid = "output-schema-invalid"
    case outputCommitRejected = "output-commit-rejected"
    case rejected
}

extension PhaseRunOutcome: Codable {
    private enum CodingKeys: String, CodingKey { case kind, code, summary, reason }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(String.self, forKey: .kind) {
        case "succeeded": self = .succeeded
        case "failed":
            self = .failed(code: try c.decode(PhaseRunFailureCode.self, forKey: .code),
                           summary: try c.decode(String.self, forKey: .summary))
        case "cancelled": self = .cancelled(reason: try c.decode(String.self, forKey: .reason))
        case let kind:
            throw DecodingError.dataCorruptedError(forKey: .kind, in: c, debugDescription: "Unknown phase run outcome kind \(kind)")
        }
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .succeeded: try c.encode("succeeded", forKey: .kind)
        case let .failed(code, summary):
            try c.encode("failed", forKey: .kind)
            try c.encode(code, forKey: .code)
            try c.encode(summary, forKey: .summary)
        case let .cancelled(reason):
            try c.encode("cancelled", forKey: .kind)
            try c.encode(reason, forKey: .reason)
        }
    }
}

/// Direct commit-binding proof for one committed output: the broker commit/tree in the project's
/// enrolled mirror and the `factory/phase/<phaseId>/<runId>` branch it published on. Deliberately
/// not `EvidenceV1`/`EvidenceSubjectV1` — that envelope's subject shape is TaskSpec/attempt-shaped
/// and has no honest values for a phase run (no TaskSpec, no policy decision, no scheduler fence).
public struct PhaseRunOutputEvidence: Hashable, Sendable, Codable {
    public var commit: GitObjectID
    public var tree: GitObjectID
    public var branch: GitBranchName

    public init(commit: GitObjectID, tree: GitObjectID, branch: GitBranchName) {
        self.commit = commit
        self.tree = tree
        self.branch = branch
    }
}

/// One committed output. `path` reuses `PhaseOutput`'s own `docs/`-rooted path shape verbatim.
public struct PhaseRunOutput: Hashable, Sendable, Codable, Identifiable {
    public var path: PhaseOutputPath
    public var digest: Sha256Digest
    public var evidence: PhaseRunOutputEvidence

    public var id: String { path.rawValue }

    public init(path: PhaseOutputPath, digest: Sha256Digest, evidence: PhaseRunOutputEvidence) {
        self.path = path
        self.digest = digest
        self.evidence = evidence
    }
}

/// A phase grader judges prose outputs against `rules.acceptanceChecks`, not source-code diffs, so
/// findings are plain bounded text — no `findingId`/`ruleId`/`category`/`severity`/`locations` like
/// `ReviewReport`'s full code-review shape. `verdict` is only `pass|changesRequired`: a phase grader
/// either accepts the produced outputs or does not.
public struct PhaseRunGraderVerdict: Hashable, Sendable, Codable {
    public enum Verdict: String, Hashable, Sendable, Codable, CaseIterable {
        case pass
        case changesRequired = "changes-required"
    }

    public var verdict: Verdict
    public var findings: [String]

    public init(verdict: Verdict, findings: [String]) {
        self.verdict = verdict
        self.findings = findings
    }
}

public struct PhaseRunTokenUsage: Hashable, Sendable, Codable {
    public var totalTokens: Int
    public init(totalTokens: Int) { self.totalTokens = totalTokens }
}

/// The durable, attempt-shaped record of one execution of a `PhaseDefinition`.
public struct PhaseRun: Hashable, Sendable, Codable, Identifiable {
    public var schemaVersion: SchemaVersion1 = .init()
    public var phaseRunId: PhaseRunID
    /// The preset this run was launched through, when one was named; `nil` for a standalone phase.
    public var presetId: PhasePresetId?
    public var phaseId: PhaseId
    public var projectId: ProjectID
    public var phaseSnapshotDigest: Sha256Digest
    public var phaseSnapshot: PhaseDefinition
    public var state: PhaseRunState
    /// Optimistic-concurrency token for this run's own state-machine writes — not a document
    /// revision like `PhaseDefinition.revision`.
    public var revision: Int
    /// The persistent room a `chat`-mode run is bound to; always `nil` for every other mode.
    public var roomId: RoomID?
    public var outputs: [PhaseRunOutput]
    public var graderVerdict: PhaseRunGraderVerdict?
    public var tokenUsage: PhaseRunTokenUsage
    public var outcome: PhaseRunOutcome?
    public var createdAt: IsoInstant
    public var startedAt: IsoInstant?
    public var finishedAt: IsoInstant?
    public var updatedAt: IsoInstant

    public var id: PhaseRunID { phaseRunId }

    /// True when a human decision (`phase.approve`/`phase.reject`) is what this run is waiting on.
    public var isAwaitingHuman: Bool { state == .awaitingHuman }

    public init(phaseRunId: PhaseRunID, presetId: PhasePresetId?, phaseId: PhaseId, projectId: ProjectID,
                phaseSnapshotDigest: Sha256Digest, phaseSnapshot: PhaseDefinition, state: PhaseRunState, revision: Int,
                roomId: RoomID?, outputs: [PhaseRunOutput], graderVerdict: PhaseRunGraderVerdict?,
                tokenUsage: PhaseRunTokenUsage, outcome: PhaseRunOutcome?, createdAt: IsoInstant, startedAt: IsoInstant?,
                finishedAt: IsoInstant?, updatedAt: IsoInstant) {
        self.phaseRunId = phaseRunId
        self.presetId = presetId
        self.phaseId = phaseId
        self.projectId = projectId
        self.phaseSnapshotDigest = phaseSnapshotDigest
        self.phaseSnapshot = phaseSnapshot
        self.state = state
        self.revision = revision
        self.roomId = roomId
        self.outputs = outputs
        self.graderVerdict = graderVerdict
        self.tokenUsage = tokenUsage
        self.outcome = outcome
        self.createdAt = createdAt
        self.startedAt = startedAt
        self.finishedAt = finishedAt
        self.updatedAt = updatedAt
    }

    private enum CodingKeys: String, CodingKey {
        case schemaVersion, phaseRunId, presetId, phaseId, projectId, phaseSnapshotDigest, phaseSnapshot, state,
             revision, roomId, outputs, graderVerdict, tokenUsage, outcome, createdAt, startedAt, finishedAt, updatedAt
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(schemaVersion, forKey: .schemaVersion)
        try c.encode(phaseRunId, forKey: .phaseRunId)
        try c.encode(presetId, forKey: .presetId)
        try c.encode(phaseId, forKey: .phaseId)
        try c.encode(projectId, forKey: .projectId)
        try c.encode(phaseSnapshotDigest, forKey: .phaseSnapshotDigest)
        try c.encode(phaseSnapshot, forKey: .phaseSnapshot)
        try c.encode(state, forKey: .state)
        try c.encode(revision, forKey: .revision)
        try c.encode(roomId, forKey: .roomId)
        try c.encode(outputs, forKey: .outputs)
        try c.encode(graderVerdict, forKey: .graderVerdict)
        try c.encode(tokenUsage, forKey: .tokenUsage)
        try c.encode(outcome, forKey: .outcome)
        try c.encode(createdAt, forKey: .createdAt)
        try c.encode(startedAt, forKey: .startedAt)
        try c.encode(finishedAt, forKey: .finishedAt)
        try c.encode(updatedAt, forKey: .updatedAt)
    }
}

/// The wire input for `phase.run`: which phase to run, for which project, from which preset (if any).
public struct PhaseRunCreate: Hashable, Sendable, Codable {
    public var presetId: PhasePresetId?
    public var phaseId: PhaseId
    public var projectId: ProjectID
    /// Overrides the phase definition's own declared `inputs[]` for this run only, when non-nil.
    public var inputsOverride: [PhaseInputKind]?

    public init(presetId: PhasePresetId?, phaseId: PhaseId, projectId: ProjectID, inputsOverride: [PhaseInputKind]?) {
        self.presetId = presetId
        self.phaseId = phaseId
        self.projectId = projectId
        self.inputsOverride = inputsOverride
    }

    private enum CodingKeys: String, CodingKey { case presetId, phaseId, projectId, inputsOverride }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(presetId, forKey: .presetId)
        try c.encode(phaseId, forKey: .phaseId)
        try c.encode(projectId, forKey: .projectId)
        try c.encode(inputsOverride, forKey: .inputsOverride)
    }
}

public struct PhaseRunListCursor: Hashable, Sendable, Codable {
    public var updatedAt: IsoInstant
    public var phaseRunId: PhaseRunID
    public init(updatedAt: IsoInstant, phaseRunId: PhaseRunID) {
        self.updatedAt = updatedAt
        self.phaseRunId = phaseRunId
    }
}

public struct PhaseRunListQuery: Hashable, Sendable, Codable {
    public var projectId: ProjectID?
    public var state: PhaseRunState?
    public var after: PhaseRunListCursor?
    public var limit: Int

    public init(projectId: ProjectID? = nil, state: PhaseRunState? = nil, after: PhaseRunListCursor? = nil, limit: Int = 100) {
        self.projectId = projectId
        self.state = state
        self.after = after
        self.limit = limit
    }

    private enum CodingKeys: String, CodingKey { case projectId, state, after, limit }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(projectId, forKey: .projectId)
        try c.encode(state, forKey: .state)
        try c.encode(after, forKey: .after)
        try c.encode(limit, forKey: .limit)
    }
}

public struct PhaseRunListPage: Hashable, Sendable, Codable {
    public var runs: [PhaseRun]
    public var nextAfter: PhaseRunListCursor?
    public var hasMore: Bool

    public init(runs: [PhaseRun], nextAfter: PhaseRunListCursor?, hasMore: Bool) {
        self.runs = runs
        self.nextAfter = nextAfter
        self.hasMore = hasMore
    }
}

/// `phase.approve`/`phase.reject` payload: which run, and why (reject only; approve reason is optional).
public struct PhaseRunDecision: Hashable, Sendable, Codable {
    public var phaseRunId: PhaseRunID
    public var reason: String?

    public init(phaseRunId: PhaseRunID, reason: String?) {
        self.phaseRunId = phaseRunId
        self.reason = reason
    }

    private enum CodingKeys: String, CodingKey { case phaseRunId, reason }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(phaseRunId, forKey: .phaseRunId)
        try c.encode(reason, forKey: .reason)
    }
}

// MARK: Command results

public struct PhaseRunResult: Hashable, Sendable, Codable {
    public var run: PhaseRun
    public init(run: PhaseRun) { self.run = run }
}
